lib/Twig/Error.php

```diff
@@ -51,13 +54,23 @@ class Twig_Error extends Exception
      *
      * By default, automatic guessing is enabled.
      *
-     * @param string    $message  The error message
-     * @param int       $lineno   The template line where the error occurred
-     * @param string    $name     The template logical name where the error occurred
-     * @param Exception $previous The previous exception
+     * @param string      $message  The error message
+     * @param int         $lineno   The template line where the error occurred
+     * @param Twig_Source $source   The source context where the error occurred
+     * @param Exception   $previous The previous exception
      */
-    public function __construct($message, $lineno = -1, $name = null, Exception $previous = null)
+    public function __construct($message, $lineno = -1, $source = null, Exception $previous = null)
     {
+        if (null === $source) {
+            $name = null;
+        } elseif (!$source instanceof Twig_Source) {
+            // for compat with the Twig C ext., passing the template name as string is accepted
```

Comment:
for BC too.

And should we update the C ext to make it call ``getSourceContext`` instead of ``getTemplateName`` ? I would say yes (should be an easy change btw, even for you or me), and then we can deprecate this case